File: public/templates/node-typescript-rag/src/pipeline/chunker.ts
/**
 * Document chunking with stable IDs and metadata.
 */

import { createHash } from "crypto";
import type { Chunk } from "../types.js";

interface ChunkOptions {
  chunkSize?: number;
  overlap?: number;
  metadata?: Record<string, unknown>;
}

/**
 * Generate a stable, reproducible chunk ID.
 */
function generateChunkId(content: string, source: string, index: number): string {
  const sourceHash = createHash("md5").update(source).digest("hex").slice(0, 8);
  const contentHash = createHash("md5").update(content).digest("hex").slice(0, 8);
  return `${sourceHash}_${contentHash}_${index.toString().padStart(4, "0")}`;
}

/**
 * Chunk a document using recursive character splitting.
 */
export function chunkDocument(
  content: string,
  source: string,
  options: ChunkOptions = {}
): Chunk[] {
  const { chunkSize = 512, overlap = 50, metadata = {} } = options;

  if (!content.trim()) {
    return [];
  }

  const chunks = recursiveSplit(content, chunkSize, overlap);

  return chunks.map((text, index) => ({
    id: generateChunkId(text, source, index),
    text,
    source,
    index,
    metadata,
  }));
}

/**
 * Recursively split text using appropriate separators.
 */
function recursiveSplit(
  text: string,
  chunkSize: number,
  overlap: number
): string[] {
  const separators = ["\n\n", "\n", ". ", ", ", " ", ""];

  if (text.length <= chunkSize) {
    return text.trim() ? [text.trim()] : [];
  }

  for (const sep of separators) {
    if (sep === "" || text.includes(sep)) {
      const splits = sep ? text.split(sep) : Array.from(text);
      const chunks: string[] = [];
      let currentChunk = "";

      for (const split of splits) {
        const testChunk = currentChunk + (currentChunk ? sep : "") + split;

        if (testChunk.length <= chunkSize) {
          currentChunk = testChunk;
        } else {
          if (currentChunk.trim()) {
            chunks.push(currentChunk.trim());
          }

          if (split.length > chunkSize) {
            // Recurse with next separator
            const sepIndex = separators.indexOf(sep);
            const subChunks = recursiveSplit(split, chunkSize, overlap);
            chunks.push(...subChunks);
            currentChunk = "";
          } else {
            currentChunk = split;
          }
        }
      }

      if (currentChunk.trim()) {
        chunks.push(currentChunk.trim());
      }

      // Add overlap
      if (overlap > 0 && chunks.length > 1) {
        return addOverlap(chunks, overlap);
      }

      return chunks;
    }
  }

  // Fallback: character-level splitting
  const result: string[] = [];
  for (let i = 0; i < text.length; i += chunkSize - overlap) {
    result.push(text.slice(i, i + chunkSize));
  }
  return result;
}

/**
 * Add overlap from previous chunk to each chunk.
 */
function addOverlap(chunks: string[], overlap: number): string[] {
  const result = [chunks[0]];

  for (let i = 1; i < chunks.length; i++) {
    const prev = chunks[i - 1];
    const prevSuffix = prev.length >= overlap ? prev.slice(-overlap) : prev;
    result.push(prevSuffix + " " + chunks[i]);
  }

  return result;
}

/**
 * Chunk markdown with header awareness.
 */
export function chunkMarkdown(
  content: string,
  source: string,
  options: ChunkOptions = {}
): Chunk[] {
  const headerPattern = /^(#{1,6})\s+(.+)$/gm;
  const sections: Array<{ header: string; content: string }> = [];

  let currentHeader = "";
  let currentContent: string[] = [];
  let lastIndex = 0;

  let match;
  while ((match = headerPattern.exec(content)) !== null) {
    // Save previous section
    const beforeHeader = content.slice(lastIndex, match.index).trim();
    if (beforeHeader || currentContent.length > 0) {
      currentContent.push(beforeHeader);
      if (currentContent.join("\n").trim()) {
        sections.push({
          header: currentHeader,
          content: currentContent.join("\n").trim(),
        });
      }
    }

    currentHeader = match[0];
    currentContent = [];
    lastIndex = match.index + match[0].length;
  }

  // Handle remaining content
  const remaining = content.slice(lastIndex).trim();
  if (remaining) {
    currentContent.push(remaining);
  }
  if (currentContent.join("\n").trim()) {
    sections.push({
      header: currentHeader,
      content: currentContent.join("\n").trim(),
    });
  }

  // Chunk each section
  const allChunks: Chunk[] = [];
  for (const section of sections) {
    const sectionText = section.header
      ? `${section.header}\n\n${section.content}`
      : section.content;

    const chunks = chunkDocument(sectionText, source, {
      ...options,
      metadata: {
        ...options.metadata,
        sectionHeader: section.header,
      },
    });

    allChunks.push(...chunks);
  }

  return allChunks;
}

File: public/templates/python-fastapi-rag/pipeline/chunker.py
"""
Document chunking with stable IDs and metadata.
"""

import hashlib
import re
from typing import Optional


def generate_chunk_id(content: str, source: str, index: int) -> str:
    """
    Generate a stable, reproducible chunk ID.
    
    Format: {source_hash}_{content_hash}_{index}
    
    This ensures:
    - Same content always gets same ID (reproducible)
    - Re-ingestion doesn't create duplicates
    - IDs are traceable to source
    """
    source_hash = hashlib.md5(source.encode()).hexdigest()[:8]
    content_hash = hashlib.md5(content.encode()).hexdigest()[:8]
    return f"{source_hash}_{content_hash}_{index:04d}"


def chunk_document(
    content: str,
    source: str,
    chunk_size: int = 512,
    overlap: int = 50,
    metadata: Optional[dict] = None,
) -> list[dict]:
    """
    Chunk a document using recursive character splitting.
    
    Args:
        content: Document text
        source: Source identifier (filename, URL, etc.)
        chunk_size: Target chunk size in characters (not tokens for simplicity)
        overlap: Overlap between chunks
        metadata: Additional metadata to attach to each chunk
    
    Returns:
        List of chunk dictionaries with text, id, and metadata
    """
    if not content.strip():
        return []
    
    # Recursive separators (most to least specific)
    separators = [
        "\n\n",      # Paragraphs
        "\n",        # Lines
        ". ",        # Sentences
        ", ",        # Clauses
        " ",         # Words
        "",          # Characters (fallback)
    ]
    
    chunks = _recursive_split(content, separators, chunk_size, overlap)
    
    result = []
    for i, chunk_text in enumerate(chunks):
        chunk_id = generate_chunk_id(chunk_text, source, i)
        result.append({
            "id": chunk_id,
            "text": chunk_text,
            "source": source,
            "index": i,
            "metadata": metadata or {},
        })
    
    return result


def _recursive_split(
    text: str,
    separators: list[str],
    chunk_size: int,
    overlap: int,
) -> list[str]:
    """Recursively split text using the most appropriate separator."""
    if len(text) <= chunk_size:
        return [text.strip()] if text.strip() else []
    
    # Find the best separator that produces chunks
    for sep in separators:
        if sep in text:
            splits = text.split(sep)
            chunks = []
            current_chunk = ""
            
            for split in splits:
                test_chunk = current_chunk + (sep if current_chunk else "") + split
                
                if len(test_chunk) <= chunk_size:
                    current_chunk = test_chunk
                else:
                    if current_chunk:
                        chunks.append(current_chunk.strip())
                    
                    # Handle splits larger than chunk_size
                    if len(split) > chunk_size:
                        # Recurse with next separator
                        sub_chunks = _recursive_split(
                            split,
                            separators[separators.index(sep) + 1:],
                            chunk_size,
                            overlap,
                        )
                        chunks.extend(sub_chunks)
                        current_chunk = ""
                    else:
                        current_chunk = split
            
            if current_chunk.strip():
                chunks.append(current_chunk.strip())
            
            # Add overlap between chunks
            if overlap > 0 and len(chunks) > 1:
                chunks = _add_overlap(chunks, overlap)
            
            return chunks
    
    # Fallback: character-level splitting
    return [text[i:i + chunk_size] for i in range(0, len(text), chunk_size - overlap)]


def _add_overlap(chunks: list[str], overlap: int) -> list[str]:
    """Add overlap from previous chunk to each chunk."""
    result = [chunks[0]]
    for i in range(1, len(chunks)):
        prev_suffix = chunks[i - 1][-overlap:] if len(chunks[i - 1]) >= overlap else chunks[i - 1]
        result.append(prev_suffix + " " + chunks[i])
    return result


# ─────────────────────────────────────────────────────────────
# Specialized chunkers
# ─────────────────────────────────────────────────────────────


def chunk_markdown(content: str, source: str, metadata: Optional[dict] = None) -> list[dict]:
    """
    Chunk markdown with header awareness.
    Preserves section context by including headers in chunks.
    """
    # Split by headers
    header_pattern = r'^(#{1,6})\s+(.+)$'
    sections = []
    current_header = ""
    current_content = []
    
    for line in content.split("\n"):
        match = re.match(header_pattern, line)
        if match:
            if current_content:
                sections.append({
                    "header": current_header,
                    "content": "\n".join(current_content),
                })
            current_header = line
            current_content = []
        else:
            current_content.append(line)
    
    if current_content:
        sections.append({
            "header": current_header,
            "content": "\n".join(current_content),
        })
    
    # Chunk each section, prepending header for context
    all_chunks = []
    for section in sections:
        section_text = f"{section['header']}\n\n{section['content']}" if section['header'] else section['content']
        chunks = chunk_document(
            section_text,
            source,
            metadata={**(metadata or {}), "section_header": section['header']},
        )
        all_chunks.extend(chunks)
    
    return all_chunks


def chunk_code(content: str, source: str, language: str = "python", metadata: Optional[dict] = None) -> list[dict]:
    """
    Chunk code files by logical boundaries (functions, classes).
    """
    # Simplified: split by function/class definitions
    if language == "python":
        pattern = r'^(def |class |async def )'
    elif language in ["javascript", "typescript"]:
        pattern = r'^(function |class |const .* = |export )'
    else:
        pattern = r'^(function |class |def )'
    
    blocks = re.split(f'(?m)(?={pattern})', content)
    
    chunks = []
    for i, block in enumerate(blocks):
        if block.strip():
            chunk_id = generate_chunk_id(block, source, i)
            chunks.append({
                "id": chunk_id,
                "text": block.strip(),
                "source": source,
                "index": i,
                "metadata": {**(metadata or {}), "language": language},
            })
    
    return chunks
